integration_tests/conftest.py

```diff
@@ -74,3 +81,335 @@ def pipeline_columns(sim, populations) -> List[str]:
+
+    This is a class so that diagnostics for an entire test run can be tracked,
+    and output to a file at the end of the run.
+    """
+
+    def __init__(self) -> None:
+        self.proportion_test_diagnostics = []
+
+    def fuzzy_assert_proportion(
+        self,
+        name: str,
+        observed_numerator: int,
+        observed_denominator: int,
+        target_proportion: Union[Tuple[float, float], float],
+        fail_bayes_factor_cutoff: float = 100.0,
+        inconclusive_bayes_factor_cutoff: float = 0.1,
+        bug_issue_beta_distribution_parameters: Tuple[float, float] = (0.5, 0.5),
+        name_additional: str = "",
+    ) -> None:
+        """
+        Assert that an observed proportion of events came from a target distribution
+        of proportions.
+        This method performs a Bayesian hypothesis test between beta-binomial
+        distributions based on the target (no bug/issue) and a "bug/issue" distribution
+        and raises an AssertionError if the test decisively favors the "bug/issue" distribution.
+        It warns, but does not fail, if the test is not conclusive (which usually
+        means a larger population size is needed for a conclusive result),
+        and gives an additional warning if the test could *never* be conclusive at this sample size.
+
+        See more detail about the statistics used here:
+        https://vivarium-research.readthedocs.io/en/latest/model_design/vivarium_features/automated_v_and_v/index.html#proportions-and-rates
+
+        :param name:
+            The name of the assertion, for use in messages and diagnostics.
+            All assertions with the same name will output identical warning messages,
+            which means pytest will aggregate those warnings.
+        :param observed_numerator:
+            The observed number of events.
+        :param observed_denominator:
+            The number of opportunities there were for an event to be observed.
+        :param target_proportion:
+            What the proportion of events / opportunities *should* be if there is no bug/issue
+            in the simulation, as the number of opportunities goes to infinity.
+            If this parameter is a tuple of two floats, they are interpreted as the 2.5th percentile
+            and the 97.5th percentile of the uncertainty interval about this value.
+            If this parameter is a single float, it is interpreted as an exact value (no uncertainty).
+            Setting this target distribution is a research task; there is much more guidance on
+            doing so at https://vivarium-research.readthedocs.io/en/latest/model_design/vivarium_features/automated_v_and_v/index.html#interpreting-the-hypotheses
+        :param fail_bayes_factor_cutoff:
+            The Bayes factor above which a hypothesis test is considered to favor a bug/issue so strongly
```

Comment:
I think it's mostly a matter of taste in this application, but any chance I could convince you to express the Bayes factor on a decibel scale, a la Jaynes and Good?

Reply:
Interesting, TIL! I am not opposed, but would prefer to postpone that to a different PR. There are a number of directions we could go here, and the one that feels most useful to me would be to specify desired sensitivity or specificity, given the prior distributions.

The intention here is that these cutoffs wouldn't be changed often. In fact, I don't think they've had to be tweaked at all here or in pseudopeople.

Reply:
yeah, i'm thinking more about the output message of a particular test than how to specify the cutoffs--regardless of how you specify the cutoff, would you report the test outcome in terms of odds like the Bayes Factor against a cutoff value? If so, reporting that in terms of decibel / log odds is nice/more interpretable for the human reading it (or at least for me 😄).

w.r.t. choosing the cutoffs, have you thought about going a step further in the default case and choose boundaries that given the prior optimize some discrimination metric like F score or log loss or something? Of course, in general I think you're right that it makes sense to let the test designer prioritize sensitivity / specificity/ some confusion matrix metric.

Reply:
Interesting ideas. I agree the output message could be improved; right now the Bayes factor gets printed in scientific notation a lot since it can get so large.

I found [this](https://www.johndcook.com/blog/2008/03/11/how-loud-is-the-evidence/); is this the same decibel scale you are thinking of? So with this cutoff, the test would fail when the evidence was more than 20 dB. Seems like a more intuitive way to think about it.

Really interesting idea about the discrimination metric. I am thinking about this in terms of decision theory, and I think in most cases the cost of a miss will be quite asymmetric. It might be interesting for _that_ to be the configurable parameter: what is the cost of not catching an issue vs raising a false alarm? But, the more fancy we get, the more we are relying on the prior distributions being somewhat accurate.

Reply:
Yes, that's right--Jaynes calls it "evidence" but it has [several names](https://en.wikipedia.org/wiki/Hartley_(unit)). The particular points in favor, one of which the post you found covers:

- It's symmetric around zero (50%), so e.g. 10 and -10 dB represent the same odds for and against
- It is additive, so it is easy to keep a running total of dB with new pieces of evidence, which proponents identify with the procedure of incremental Bayesian updates
- Folks who are familiar with working with a decibel scale may know the simple power ratio heuristic 3 dB ~ 2x power and 10 dB = 10x power
-  It is speculated to be calibrated to human perception in the vein of the [Weber Fetchner law](https://en.wikipedia.org/wiki/Weber%E2%80%93Fechner_law), so that 1 dB is about as much "evidential strength" that a person can meaningfully distinguish (~25% increase)

> I am thinking about this in terms of decision theory, and I think in most cases the cost of a miss will be quite asymmetric. It might be interesting for that to be the configurable parameter: what is the cost of not catching an issue vs raising a false alarm?

Yeah I'm not super familiar with evaluation metrics--there seem to be many, especially for binary classifiers--but just looking at wikipedia they show a [parameterized Beta f-score](https://en.wikipedia.org/wiki/F-score) that weights the harmonic mean toward precision or recall. I wonder if it would be sensible to do something similar with sensitivity / specificity. At the end of the day you're trying to trade off type I and type II errors i guess, but I'm not sure how it all pencils out. But it would be nice to say "I want to prioritize sensitivty / specificity X times more than the other" without having to know the right cutoffs

> But, the more fancy we get, the more we are relying on the prior distributions being somewhat accurate.
Yeah, this is definitely an argument not to go too crazy with the information theory stuff

Reply:
I believe any decision metric will correspond directly to some loss function. In the case of a binary classification, the loss function is just a 2x2 matrix of how good/bad it is to have any combination of predicted and true class. My guess is that the loss function itself will be more intuitive to specify than the corresponding metric or even the desired sensitivity/specificity.